### ISSUE TITLE
asyncio.gather() adds significant overhead

### ISSUE THREAD

**Issue by tyoung-patreon** [strawberry-graphql/strawberry#2234]
We're in the process of migrating from https://github.com/quora/asynq to asyncio using dataloaders. Without getting too into the details, asynq provides the same general premise of batching and caching as dataloaders do, but it uses python generators and a custom scheduler to achieve it over the asyncio event loop.

In our fairly large application switching to asyncio dataloaders adds significant overhead (multiple seconds of e2e latency), and this comes down to the fact that the overhead of asyncio.gather is significant.

Consider the following trivialized example:

```python
import asyncio
import dataclasses
import time
from strawberry.dataloader import DataLoader


@dataclasses.dataclass(frozen=True)
class Result:
    x: int = 0


async def load_fn(keys):
    await asyncio.sleep(0.1)  # Synthesize network latency
    return [Result(i) for i in keys]


async def non_gather_version(size, loader):
    awaitables = []
    for i in range(size):
        awaitables.append(loader.load(i))
    result = []
    for a in awaitables:
        result.append(await a)
    return result


def main():
    r = 100000
    loader = DataLoader(load_fn=load_fn, cache=False)
    loop = asyncio.get_event_loop()

    t0 = time.time()
    loop.run_until_complete(loader.load_many(range(r)))
    print("Asyncio Gather Version:", time.time() - t0)

    t0 = time.time()
    loop.run_until_complete(non_gather_version(r, loader))
    print("Non Gather Version:", time.time() - t0)

    t0 = time.time()
    loop.run_until_complete(load_fn(range(r)))
    print("Direct Version:", time.time() - t0)

if __name__ == "__main__":
    main()
```

On my modest aws machine I get the following results:

```
Asyncio Gather Version: 1.255368947982788
Non Gather Version: 0.6231565475463867
Direct Version: 0.1942920684814453
```

The overhead from gather is readily apparent, and we have experienced significantly worse in our non trivial workloads. Our codebase is designed around these wide and deep call hierarchy fan-outs where we use multiple dataloaders at various depths in our call stack. At each fanout point we consolidate the child awaitables with a call to gather. When this is scaled up to the hundreds of thousands of objects (mostly duplicated or cached with dataloaders), we incur a performance penalty we cannot bear.

I don't understand how to use dataloaders in a performant fashion, it seems for them to be useful it necessitates that they are available at multiple levels of the call stack (else you would just make a single batched call directly yourself), but this also makes them very expensive, compared to both calling batches directly or using the generator approach the referenced asynq uses above.

**Comment by ThirVondukr** [comment]
Dataloaders mostly used in a different way, if you need to load a lot of objects in single place - just do it directly.
They're are great when it comes to loading related items/small collections for a lot of objects, for example like this:

```py
@strawberry.type(name="User")
class UserType:
    id: strawberry.ID
    name: str


async def load_user_by_id(user_ids: list[UUID]) -> list[UserType]:
    stmt = select(User).where(User.id.in_(filter_uuids(user_ids)))

    async with async_sessionmaker() as session:
        users = {user.id: user for user in await session.scalars(stmt)}

    return [users.get(str(id)) for id in user_ids]


@strawberry.type(name="Book")
class BookType:
    title: str
    author_id: UUID

    @strawberry.field
    async def author(self, info: Info) -> UserType:
        return await info.context.loaders.user_by_id.load(self.author_id)
```

**Comment by tyoung-patreon** [comment]
Hi @ThirVondukr - I understand that a batch invocation through dataloader.load_many is not usual practice. That was only used to describe the generality of the problem. Here is a less trivial demonstration of the principle at play; 

```python
async def get_account_balance(user_name) -> int:
    user = await user_data_loader.load(user_name)
    account_list = await account_data_loader.load(user.user_id)
    account_balances = await balance_data_loader.load_many(account_list)
    return sum(account_balances)
```

Hopefully this demonstrates the fan-out I mention and necessity of `gather` (aka load_many). I realize this example is still fairly trivial, but we can imagine things happening in between the lines that would make it difficult to refactor. 

With the call pattern `asyncio.gather(*[get_account_balance(name) for name in list_of_names])` and the power of dataloaders we have correctly only made three queries to the database (woohoo), but it has come at the significant overhead of adding a task to the event loop for each name, and then later for each account. This overhead is expensive (as demonstrated above).

It seems to me this is the entire point of dataloaders, so that I can fan-out to multiple helper functions which each only call load and do some operation on the result. But in order to have that fan out over a list (so that they are batched), it requires I `asyncio.gather()` on that function. If the example above did not use `load_many` it would look like:

```python
    for account in account_list:
        await balance_data_loader.load(account)
```

and that would defeat the purpose of dataloaders, since each call would not be batched! Sure here I could do something like:

```python
    futures = [balance_data_loader.load(account) for account in account_list]
    result = [await future for future in futures]
```

to avoid a `gather()` call, but then I can't use a helper function that does additional work and itself calls `await balance_data_loader.load(account)`. So in the case I want to use a helper function, I have to use `gather()` to make the results useful.

Essentially every fan-out operation is going to require a call to asyncio.gather() or it is going to be extremely unergonomic. And asyncio.gather() is expensive because it puts every single argument onto the event loop. For a non-trivial application with hundreds of thousands of calls into dataloaders, this makes all of them extremely expensive, even if it is only executing a handful of queries. Even the caching feature of dataloders will be put onto the event loop!

**Comment by ThirVondukr** [comment]
Did you try running your application with [uvlop](https://github.com/MagicStack/uvloop) and see if it makes a difference? Just curious.
For that aggregation you could write separate dataloader that loads your balance sum by user id:
```py
@strawberry.type(name="User")
class UserType:
    _instance: Private[User]
    id: strawberry.ID
    name: str

    async def balance_sum(self, info: Info) -> int:
        return info.context.loaders.user_balance.load(self._instance.id)


async def load_user_balance_sum(user_ids: list[UUID]) -> list[UserType]:
    stmt = (
        select(User, func.sum(Wallet.balance))
        .join(User.wallet)
        .where(User.id.in_(user_ids))
        .group_by(User.id)
    )

    async with async_sessionmaker() as session:
        balances = {
            user.id: balance_sum
            for user, balance_sum in await session.execute(stmt)
        }

    return [balances.get(id, 0) for id in user_ids]
```

**Comment by tyoung-patreon** [comment]
Yeah I had tried uvloop to no measurable improvement

**Comment by patrick91** [comment]
I'll close this since it is quite old, and hasn't been brought up for a while 😊

**Comment by mgasner** [comment]
This is definitely a real problem when using strawberry with real world workloads -- we are running into this same issue and actually factoring our hot query paths to not use dataloaders as a consequence.

**Comment by patrick91** [comment]
@mgasner hello! can you open a new issue with a reproduction? 

this might be more on graphql-core side, but without reproduction it's hard to check 😊
